traits/src/fee.rs

```diff
@@ -0,0 +1,28 @@
+use frame_support::sp_runtime::{DispatchError, DispatchResult};
+
+//TODO: give better name and add doc comment
+pub trait InspectSufficiency<AssetId> {
+	fn is_sufficient(asset: AssetId) -> bool;
+
+	fn is_trade_supported(from: AssetId, into: AssetId) -> bool;
+}
+
+//TODO: give better name when we decicded if we wanna hardcode DOT or not
+pub trait InsufficientAssetTrader<AccountId, AssetId, Amount>: InspectSufficiency<AssetId> {
+	fn buy(
+		origin: &AccountId,
+		dest: &AccountId,
+		from: AssetId,
+		into: AssetId,
+		amount: Amount,
+		max_limit: Amount,
+	) -> DispatchResult;
+
+	fn pool_trade_fee(swap_amount: Amount) -> Result<Amount, DispatchError>;
```

Comment:
why result is amount ?! should be Permill ? or more generic PerThing ?

Reply:
Here I am calculating the pool trade fee which is the ExchangeFeeRate multiplied by an amount.

We could return only the ExchangeFeeRate, but then the fee calculation needs to happen where this function is used, but that would require xyk specific code, such as:

[https://github.com/galacticcouncil/hydration-node/blob/6cab9f7e235c7a050e8c30163f006e565225a641/runtime/hydradx/src/assets.rs#L1755-L1760](https://github.com/galacticcouncil/hydration-node/blob/6cab9f7e235c7a050e8c30163f006e565225a641/runtime/hydradx/src/assets.rs#L1755-L1760)

Reply:
i would rename it to calculate fee amount.